resources/schema.py

```diff
@@ -170,6 +171,25 @@ def resolve_image_file(self, info, **kwargs):
             return None
 
 
+class AbstractAreaInput(graphene.InputObjectType):
+    servicemap_id = graphene.String()
+    zip_code = graphene.String()
+    phone = graphene.String()
+    email = graphene.String()
+    www_url = graphene.String()
+    location = graphql_geojson.Geometry()
+    image_link = graphene.String()
+
+
+class HarborAreaInput(AbstractAreaInput):
```

Comment:
you could probably rename it to `HarborInput` and have all the other fields added right here. so that eventually the mutation looks like this:
```
"createHarbor": {
    "municipality": "helsinki",
    "imageFile": "https://foobar.com/image.png",
    "availabilityLevel": "1",
    "numberOfPlaces": 150,
    "maximumWidth": 350,
    "maximumLength": 400,
    "maximumDepth": 100,
    "name": "Foobarsatama",
    "streetAddress": "Foobarsatamantie 1234",
    "servicemapId": "1",
    "location": {
      "type": "Point",
      "coordinates": [125.6, 10.1]
    },
    "zipCode": "00119",
    "phone": "123123123",
    "email": "foo@bar.com",
    "wwwUrl": "https://foobar.com",
    "imageLink": "https://foobar.com/image.png"
}
```

Reply:
I tried about doing something like that, either composing or extending the `AbstractAreaInput` type, but couldn't find a way to do it (not as straightforward as it sounds). That's why I decided to keep `area` as a separate input that could be reusable.

I can dedicate some more time to see if there could be another way to achieve that behavior.

Reply:
would this work? :
```python
class AbstractAreaInput(graphene.InputObjectType):
    servicemap_id = graphene.String()
    zip_code = graphene.String()
    phone = graphene.String()
    email = graphene.String()
    www_url = graphene.String()
    location = graphql_geojson.Geometry()
    image_link = graphene.String()


class HarborInput(AbstractAreaInput):
    municipality = graphene.String()
    image_file = graphene.String()
    availability_level = graphene.ID()
    number_of_places = graphene.Int()
    maximum_width = graphene.Int()
    maximum_length = graphene.Int()
    maximum_depth = graphene.Int()
    name = graphene.String()
    street_address = graphene.String()


class CreateHarborMutation(graphene.ClientIDMutation):
    class Input:
        harbor = HarborInput(required=True)
```

not exactly the way I initially planned, where all the fields are directly declared on the `Input`, but I just don't see any ontological reason to have some of the fields within the `area` field, and some directly on the `Harbor` object.

I used `AbstractArea` only to minimize the lines of code, as there were common fields between `WinterStorageArea` and `Harbor`, not because they have some `area` entity within them.

Reply:
I first thought about extending the `Input`:
```python
class CreateHarborMutation(graphene.ClientIDMutation):
    class Input(AbstractAreaInput):
        additional_fields = graphene.String()
```

But when you try to extend another `InputObjectType` from the `Input` class, it throws this error:

```
berth       | TypeError: Cannot create a consistent method resolution
berth       | order (MRO) for bases InputObjectType, Input
```

I didn't dig much into it tho. It might be doable overriding a couple of methods, but who knows. 🤷‍♂ 

--- 

Also, but doing something like
```python
class CreateHarborMutation(graphene.ClientIDMutation):
    class Input:
        harbor = HarborInput(required=True)
```
the input would have the shape:
```json
{
  "input": {
    "harbor": {
      "foo": "bar",
    }
}
```
so there would be an unnecessary additional `harbor` field, BUT it's doable.

Reply:
You can ignore all of my previous blabbering. I found out why was it failing and how to properly do it. I'll update this on Monday with tons of changes.

Reply:
Alright, looks very good the way it is now! 👍 🏅